File: preprocessing/applyCLAHE.m
function outImg = applyCLAHE(inImg, clipLimit, tileGridSize, colorMode)
% APPLYCLAHE Contrast-Limited Adaptive Histogram Equalization for fundus images.
%
%   outImg = applyCLAHE(inImg)
%   outImg = applyCLAHE(inImg, clipLimit)
%   outImg = applyCLAHE(inImg, clipLimit, tileGridSize)
%   outImg = applyCLAHE(inImg, clipLimit, tileGridSize, colorMode)
%
%   Inputs:
%       inImg        - RGB or grayscale retinal fundus image (uint8 or double).
%       clipLimit    - (Optional) Contrast threshold limit (Default: 0.02, or integer e.g. 2.0).
%       tileGridSize - (Optional) 2-element vector [M, N] contextual tiles (Default: [8, 8]).
%       colorMode    - (Optional) 'Lab' (Default) or 'GreenChannel'.
%
%   Outputs:
%       outImg       - CLAHE-enhanced RGB or grayscale image (uint8).
%
%   Clinical Rationale:
%       Enhancing the L* channel in CIELAB or the green channel selectively boosts
%       the visibility of subtle microaneurysms and intraretinal microvascular
%       abnormalities (IRMA) without colorimetric distortion.
%
%   Author: SIH26038 Capstone Engineering Team
%   Date: September 2026

    if nargin < 2 || isempty(clipLimit)
        clipLimit = 0.02;
    elseif clipLimit > 1.0
        clipLimit = clipLimit / 100.0; % Convert percentage/integer factor (e.g. 2.0 -> 0.02)
    end

    if nargin < 3 || isempty(tileGridSize)
        tileGridSize = [8, 8];
    end

    if nargin < 4 || isempty(colorMode)
        colorMode = 'Lab';
    end

    if isfloat(inImg) && max(inImg(:)) <= 1.0
        imgUint8 = im2uint8(inImg);
    else
        imgUint8 = uint8(inImg);
    end

    if size(imgUint8, 3) == 3
        switch lower(colorMode)
            case 'lab'
                % CIELAB color space: Enhance L* (luminance) channel
                labImg = rgb2lab(imgUint8);
                L_norm = labImg(:, :, 1) / 100.0;
                
                L_enh = adapthisteq(L_norm, ...
                                    'ClipLimit', clipLimit, ...
                                    'NumTiles', tileGridSize, ...
                                    'Distribution', 'rayleigh');
                
                labImg(:, :, 1) = L_enh * 100.0;
                outImg = lab2rgb(labImg);
                outImg = im2uint8(max(0.0, min(1.0, outImg)));

            case 'greenchannel'
                % Green channel selective enhancement
                outImg = imgUint8;
                outImg(:, :, 2) = adapthisteq(imgUint8(:, :, 2), ...
                                              'ClipLimit', clipLimit, ...
                                              'NumTiles', tileGridSize, ...
                                              'Distribution', 'rayleigh');

            otherwise
                % Default to Lab
                labImg = rgb2lab(imgUint8);
                L_norm = labImg(:, :, 1) / 100.0;
                L_enh = adapthisteq(L_norm, 'ClipLimit', clipLimit, 'NumTiles', tileGridSize);
                labImg(:, :, 1) = L_enh * 100.0;
                outImg = im2uint8(lab2rgb(labImg));
        end
    else
        % Grayscale direct CLAHE
        outImg = adapthisteq(imgUint8, 'ClipLimit', clipLimit, 'NumTiles', tileGridSize);
    end
end

File: preprocessing/apply_clahe.py
"""
preprocessing/apply_clahe.py
============================
Contrast Limited Adaptive Histogram Equalization (CLAHE) on L*a*b* space.
Decouples luminance from color chrominance to prevent vascular artifact distortion.
"""

import numpy as np

def apply_clahe(image: np.ndarray, clip_limit: float = 2.0, tile_grid_size: tuple = (8, 8)) -> np.ndarray:
    """
    Applies CLAHE to the L* (Lightness) channel of an RGB retinal image.
    
    Args:
        image: RGB uint8 array of shape (H, W, 3).
        clip_limit: Threshold for contrast limiting (default 2.0 / 0.02 normalized).
        tile_grid_size: Number of contextual tiles (default 8x8).
        
    Returns:
        Enhanced RGB uint8 image with amplified microaneurysms and vessel clarity.
    """
    try:
        import cv2
        lab = cv2.cvtColor(image, cv2.COLOR_RGB2LAB)
        l_channel, a_channel, b_channel = cv2.split(lab)
        
        clahe = cv2.createCLAHE(clipLimit=clip_limit, tileGridSize=tile_grid_size)
        cl = clahe.apply(l_channel)
        
        enhanced_lab = cv2.merge((cl, a_channel, b_channel))
        enhanced_rgb = cv2.cvtColor(enhanced_lab, cv2.COLOR_LAB2RGB)
        return enhanced_rgb
    except ImportError:
        # Graceful fallback: histogram stretch on luminance
        img_float = image.astype(np.float32) / 255.0
        p2, p98 = np.percentile(img_float, (2, 98))
        stretched = np.clip((img_float - p2) / max(1e-5, (p98 - p2)), 0.0, 1.0)
        return (stretched * 255).astype(np.uint8)

if __name__ == "__main__":
    sample = np.random.randint(0, 255, (224, 224, 3), dtype=np.uint8)
    out = apply_clahe(sample)
    print(f"CLAHE applied: Input {sample.shape} -> Output {out.shape}")
